Scatter-Gather part1 (#1)
timbussmann/timbussmann.github.io: _posts/2019-09-25-scatter-gather-pt1.md

---
layout: post
title: Scatter-Gather in NServiceBus, part 1
---

> This is article series is about showing how to implement the Scatter-Gather pattern correctly in NServiceBus. I've created this blog post because there are some common mistakes you can easily fall for, leading to bad performance and other issues.

Reviewer: ```suggestion
This article series is about showing how to implement the Scatter-Gather pattern correctly in NServiceBus. There are some common mistakes and traps users can easily fall into, leading to bad performance and other issues.
```

I prefer to refer to users instead of you, feel free to change it back to you if you prefer to look it like a chat between you and the reader.

> Scatter-Gather is one of the patterns described in the [Enterprise Integration Patterns book](https://www.amazon.com/Enterprise-Integration-Patterns-Designing-Deploying/dp/0321200683). The [description of the pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/BroadcastAggregate.html) states:

Reviewer: ```suggestion
Scatter-Gather is one of the patterns mentioned in the [Enterprise Integration Patterns book](https://www.amazon.com/Enterprise-Integration-Patterns-Designing-Deploying/dp/0321200683). [It is described as](https://www.enterpriseintegrationpatterns.com/patterns/messaging/BroadcastAggregate.html):
```

> Use a Scatter-Gather that broadcasts a message to multiple recipients and re-aggregates the responses back into a single message.

> The pattern consists of two parts: Broadcasting a message to multiple recipients and then using the [Aggregator pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Aggregator.html) to gather responses. A variation is to split an incoming message into smaller sub-messages (which each produce a response) instead of broadcasting. This variation is described as the [Splitter pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Sequencer.html). So technically, in this blog post, I'm going to use a Splitter+Aggregator sample but it's not a big difference to using the actual Scatter-Gather pattern when using NServiceBus and everything described here holds true in both cases.

Reviewer: ```suggestion
The pattern consists of two parts: Broadcasting a message to multiple recipients and then using the [Aggregator pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Aggregator.html) to gather responses. A variation is to split an incoming message into smaller sub-messages (each one producing a response) instead of broadcasting. This variation is described as the [Splitter pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Sequencer.html).

In this blog post, I'll use a Splitter+Aggregator sample. All what is described here holds true for the Scatter-Gather pattern, especially when using NServiceBus.
```

I’m not sure about this last change, as it’s not clear what’s the original intent.
Consider also adding a sample for each of the described patterns.

> Let's just get started:

Reviewer: ```suggestion
```

> ### Scatter:

Reviewer: ```suggestion
### Scatter
```

Scattering (or broadcasting, or splitting) is fairly easy with NServiceBus. This sample handler shows how to split an incoming message in a bunch of subsequent messages for further processing. Alternatively, this could also be a handler publishing an event to multiple subscribers.

```
public async Task Handle(OrderPlaced message, IMessageHandlerContext context)
{
    foreach (var orderItem in message.OrderItems)
    {
        await context.Send(new ProcessOrderItem
        {
            OrderId = message.OrderId,
            ItemId = orderItem.Id
        });
    }
}
```

> NServiceBus takes care of batching the outgoing messages already. If the handler fails with an exception at any point, no message is sent*.

Reviewer: ```suggestion
NServiceBus takes care of batching outgoing messages: If the handler fails with an exception at any point, no message is sent*.
```

Maybe add “and the incoming message is retried”?

> ### Gather:

Reviewer: ```suggestion
### Gather
```

> Let's say we expect a response message from every previously sent message and we want to continue the process once all responses have been received. This means we need some persistent state to keep track of the expected responses. Very often, people decide to use a [Saga](https://docs.particular.net/nservicebus/sagas) for this, as it provides a shared state across multiple message handlers. The Scatter-Gather implementation would probably look something like this:

Reviewer: ```suggestion
Let's say we expect a response message for every sent message, and we want to continue the business process once all responses have been received. This means we need some persistent state to keep track of the expected responses. Very often, people decide to use a [Saga](https://docs.particular.net/nservicebus/sagas) for this, as it provides a shared state across multiple message handlers. The Scatter-Gather implementation would probably look something like this:
```

    IAmStartedByMessages<OrderPlaced>,
    IHandleMessages<ProcessOrderItemResult>
{
    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ScatterGatherSagaData> mapper)
    {
        mapper.ConfigureMapping<OrderPlaced>(m => m.OrderId).ToSaga(s => s.OrderId);
        mapper.ConfigureMapping<ProcessOrderItemResult>(m => m.OrderId).ToSaga(s => s.OrderId);
    }

    public async Task Handle(OrderPlaced message, IMessageHandlerContext context)
    {
        Data.NumberOfItems = message.OrderItems.Count();
        foreach (var orderItem in message.OrderItems)
        {
            await context.Send(new ProcessOrderItem
            {
                OrderId = message.OrderId,
                ItemId = orderItem.Id
            });
        }
    }

    public async Task Handle(ProcessOrderItemResult message, IMessageHandlerContext context)
    {
        var totalProcessedOrders = ++Data.ProcessedOrders;
        if (totalProcessedOrders == Data.NumberOfItems)
        {
            await context.Publish(new OrderProcessed { OrderId = Data.OrderId });
            MarkAsComplete();
            Console.WriteLine("Order completed");
        }
    }
}
```

> This looks quite nice, the Saga API provides a simple solution to implement the Scatter-Gather pattern. Unfortunately, it's a very suboptimal solution and you will run into problems. 

Reviewer: ```suggestion
This looks quite nice, the Saga API provides a simple solution to implement the Scatter-Gather pattern. Unfortunately, it's a very suboptimal solution that will run into issues at runtime. 
```

### Sagas and concurrency

> Sagas are specifically designed to implement long-running business processes, while Scatter-Gather typically isn't very long-running but mostly just a short-lived aggregation process. This short-lived nature is also exactly what causes the problem with the shown implementation. In most cases, scattering is used to parallelize work which means that we expect a lot of responses to arrive back to the saga at the same time. When a handler of a saga is invoked, the data of the saga is loaded from the database. This leads to a situation where multiple responses want to update the same saga instance concurrently (keep in mind that NServiceBus can handle multiple messages concurrently by default). Sagas can't handle concurrency, there is only one message which can successfully complete a saga at one time because they all touch the same data.

Reviewer: I’m not sure I agree with the definition of long-running. I’d define long running as “spans multiple physical transactions” and not “based on the time it takes”.

```suggestion
In most cases, scattering is used to parallelize work which means that we expect a lot of responses to arrive back to the saga at the same time. When a handler of a saga is invoked, the data of the saga is loaded from the database. This leads to a situation where multiple responses want to update the same saga instance concurrently (keep in mind that NServiceBus can handle multiple messages concurrently by default). Sagas handle concurrency by making sure only one message at a time can successfully update the same data.
```

> Now, there are a few differences between the available saga persisters which boils down to whether they use pessimistic or optimistic concurrency control to resolve concurrency issues on the same saga instance.** 

Reviewer: ```suggestion
There are a few differences between the available saga persisters which boil down to whether they use pessimistic or optimistic concurrency control to resolve concurrency issues on the same saga instance.** 
```

While neither of the solutions is particularly helpful in this case, optimistic concurrency can make the situation a lot worse. Assuming that many messages to the same saga are in the queue, retrying a message due to a concurrency exception won't help much, as it will just enter a new race condition every time. After failing multiple times, recoverability will fail the message and move it to the error queue. This means the user has to manually retry the message later on instead of recoverability handling the issue automatically.

> I hope you now have a better understanding of why a Saga is a bad choice to implement the Scatter-Gather patten in almost all cases. The only exceptions might be situations where the responses trickle in slowly after each other than in parallel or when you're only scattering a very low amount of messages (< 5). Sagas are made to implement long-running processes and while their API might invite for other applications sometimes, there is a high chance that it ends up being a bad choice.

Reviewer: ```suggestion
I hope you now have a better understanding of why a NServiceBus Saga is a suboptimal choice to implement the Scatter-Gather patten in almost all cases. The only exception might be a situation where responses trickle in slowly after each other, or when you're only scattering a very low amount of messages (< 5). Sagas are made to implement long-running processes and while their API might invite for other implementations, there is a high chance that it ends up being a bad choice.
```

I personally don’t like to much the word bad.

We will look into a better solution for the Scatter-Gather pattern in the next blog post.

> *There is no absolute guarantee that the outgoing messages are atomic with completing the incoming message unless you're using a transaction mode of `SendsAtomicWithReceive` or higher. There is always some edge case that might interrupt the actual dispatch operation (e.g. a network failure in the middle of dispatching the batch), but NServiceBus does it's best to keep the window for such edge cases as small as possible. This is a topic for another blog post though.

Reviewer: Move this up, where the asterisk is as a note/quote. Also do you really want to enter in this kind of details? We could argue that NServiceBus has ways, e.g. the Outbox, to guarantee consistency. I’m not sure this note is worth it

Author: I wanted to avoid this level of technical details on a post about sagas but felt that the linked statement might give some wrong impressions. I think removing it might makes sense

> **SQL Persistence [uses pessimistic concurrency control for querying saga data](https://docs.particular.net/persistence/sql/saga-concurrency#concurrent-access-to-existing-saga-instances). NHibernate persistence [can be configured to use pessimistic concurrency control](https://docs.particular.net/persistence/nhibernate/saga-concurrency#adjusting-the-locking-strategy).

Reviewer: Move this up, where the asterisk is as a note/quote.

---
layout: post
title: Scatter-Gather in NServiceBus, part 1
---

This article series is about showing how to implement the Scatter-Gather pattern in NServiceBus. There are some common mistakes and hidden traps users can easily fall into, leading to bad performance and other issues.

Scatter-Gather is one of the patterns mentioned in the [Enterprise Integration Patterns book](https://www.amazon.com/Enterprise-Integration-Patterns-Designing-Deploying/dp/0321200683). [It is described as](https://www.enterpriseintegrationpatterns.com/patterns/messaging/BroadcastAggregate.html):

> The Scatter-Gather routes a request message to the a number of recipients. It then uses an Aggregator to collect the responses and distill them into a single response message.

The pattern consists of two parts: Broadcasting a message to multiple recipients and then using the [Aggregator pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Aggregator.html) to gather responses.

### Broadcast

> There are multiple ways to "broadcast" a message to multiple recipients. Typically, broadcasting is associated with [Publish-Subscribe](https://www.enterpriseintegrationpatterns.com/patterns/messaging/PublishSubscribeChannel.html) which is implemented using the `Publish` API in NServiceBus. However, there are many different patterns to send messages to multiple recipients and it depends on the type of the message, the recipients and the form of coupling between sender and receivers to find the best way to "broadcast".

Reviewer: ```suggestion
There are multiple ways to "broadcast" a message to multiple recipients. Typically, broadcasting is associated with [Publish-Subscribe](https://www.enterpriseintegrationpatterns.com/patterns/messaging/PublishSubscribeChannel.html) which is implemented using the `Publish` API in NServiceBus. However, there are many different ways to send messages to multiple recipients depending on the type of the message, the recipients, and the form of coupling between senders and receivers.
```

> In this blog post, I'm going to use the [Splitter pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Sequencer.html): An incoming message is split into multiple smaller messages which can then be processed independently (and parallel) from each other. The result of each "sub-message" is then aggregated back to a single result again. This pattern is implemented using regular `Send` operations in NServiceBus. Strictly speaking, this might not match 100% with the definition of the Scatter-Gather pattern as we're sending multiple messages of the same type (with different values) but for the remainer of this blog post, it doesn't make a big difference whether you publish an event or send commands.

Reviewer: ```suggestion
In this blog post, I'm going to use the [Splitter pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Sequencer.html): An incoming message is split into multiple smaller messages which can then be processed independently (and in parallel) from each other. The result of each "sub-message" is then aggregated back to a single result again. This pattern is implemented using regular `Send` operations in NServiceBus. Strictly speaking, this might not match 100% with the definition of the Scatter-Gather pattern as we're sending multiple messages of the same type (with different values) but for the remainer of this blog post, it doesn't make a big difference whether you publish an event or send commands.
```

        foreach (var orderItem in message.OrderItems)
        {
            await context.Send(new ProcessOrderItem
            {
                OrderId = message.OrderId,
                ItemId = orderItem.Id
            });
        }
    }

    public async Task Handle(ProcessOrderItemResult message, IMessageHandlerContext context)
    {
        var totalProcessedOrders = ++Data.ProcessedOrders;
        if (totalProcessedOrders == Data.NumberOfItems)
        {
            await context.Publish(new OrderProcessed { OrderId = Data.OrderId });
            MarkAsComplete();
            Console.WriteLine("Order completed");
        }
    }
    
    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ScatterGatherSagaData> mapper)
    {
        mapper.ConfigureMapping<OrderPlaced>(m => m.OrderId).ToSaga(s => s.OrderId);
        mapper.ConfigureMapping<ProcessOrderItemResult>(m => m.OrderId).ToSaga(s => s.OrderId);
    }
}
```

This looks quite nice, the Saga API provides a simple solution to implement the Scatter-Gather pattern. Unfortunately, it's a very suboptimal solution that will run into issues at runtime. 

### Sagas and concurrency

> When a handler of a saga is invoked, the state of the saga is loaded from the database. This leads to a situation where multiple responses want to update the same saga instance concurrently (keep in mind that NServiceBus can handle multiple messages concurrently by default). Sagas can't handle concurrency, there is only one message which can successfully complete a saga at one time because they all touch the same data.

Reviewer: “Can’t handle” reads as a limitation. As something that cannot be done. I guess that what you want to say is that sagas are designed o handle concurrency in such a way that it doesn’t result in an inconsistent state.

Author: @mauroservienti what do you think about 
> Sagas can't resolve concurrency, there is only one message which can successfully complete a saga at one time because they all touch the same data.

?

> There are a few differences between the available saga persisters which boil down to whether they use pessimistic or optimistic concurrency control to resolve concurrency issues on the same saga instance.**

Reviewer: ```suggestion
There are a few differences between the available saga persisters which boil down to whether they use pessimistic or optimistic concurrency control to resolve concurrency issues on the same saga instance.
```

* With optimistic concurrency control, concurrent access to the same saga data is only noticed once the message handler has completed and tries to update the saga's state. At this point, a concurrency exception will be raised by the database, failing the message. Recoverability mechanism will then ensure that the message is retried. You will notice some sort of concurrency related exceptions in your endpoint's log files if that occurs. Most supported NServiceBus persisters use optimistic concurrency control.

> * With pessimistic concurrency control, concurrent attempts to load the same sagas are blocked (by using some form of locks) to ensure sequential access to the data. This means that there will most likely be no exceptions visible. However, the endpoint message processing throughput will slow down significantly as in the worst case only one message can be processed at the same time. This means that with pessimistic concurrency control, the endpoint can fall down to a concurrency level of one, allowing only a single message to complete after another. This can potentially cause database servers to assume a deadlock if the locks are held for too long, in which case you'd also see exceptions showing up. SQL Persistence [uses pessimistic concurrency control for querying saga data](https://docs.particular.net/persistence/sql/saga-concurrency#concurrent-access-to-existing-saga-instances) and NHibernate persistence [can be configured to use pessimistic concurrency control](https://docs.particular.net/persistence/nhibernate/saga-concurrency#adjusting-the-locking-strategy).

Reviewer: ```suggestion
* With pessimistic concurrency control, concurrent attempts to load the same sagas are blocked (by using some form of locks) to ensure sequential access to the data. This means that there will most likely be no exceptions visible. However, the endpoint message processing throughput will slow down significantly as in the worst case only one message can be processed at the same time. This means that with pessimistic concurrency control, the endpoint can fall down to a concurrency level of one, allowing only a single message to complete after another. This can potentially cause database servers to assume a deadlock if the locks are held for too long, in which case you'd also see exceptions showing up. SQL Persistence, for example, [uses pessimistic concurrency control for querying saga data](https://docs.particular.net/persistence/sql/saga-concurrency#concurrent-access-to-existing-saga-instances) and NHibernate persistence [can be configured to use pessimistic concurrency control](https://docs.particular.net/persistence/nhibernate/saga-concurrency#adjusting-the-locking-strategy).
```
